src/lifo.zig

```diff
@@ -0,0 +1,149 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+const constants = @import("./constants.zig");
+
+/// An intrusive last in/first out (LIFO, stack) linked list.
+/// The element type T must have a field called "next" of type ?*T.
+pub fn LIFOType(comptime T: type) type {
+    return struct {
+        const LIFO = @This();
+
+        head: ?*T = null,
+        count: u64 = 0,
+
+        // This should only be null if you're sure we'll never want to monitor `count`.
+        name: ?[]const u8,
+
+        // If the number of elements is large, the constants.verify check in push() can be too
+        // expensive. Allow the user to gate it. Could also be a comptime param?
+        verify_push: bool = true,
+
+        pub fn push(self: *LIFO, elem: *T) void {
+            if (constants.verify and self.verify_push) assert(!self.contains(elem));
+            assert(elem.next == null);
+            // Insert the new element at the head.
+            elem.next = self.head;
+            self.head = elem;
+            self.count += 1;
+        }
+
+        pub fn pop(self: *LIFO) ?*T {
+            const ret = self.head orelse return null;
+            self.head = ret.next;
+            ret.next = null;
+            self.count -= 1;
+            return ret;
+        }
+
+        pub fn peek(self: LIFO) ?*T {
+            return self.head;
+        }
+
+        pub fn empty(self: LIFO) bool {
+            return self.head == null;
+        }
+
+        /// Returns whether the linked list contains the given *exact element* (pointer comparison).
+        pub fn contains(self: *const LIFO, elem_needle: *const T) bool {
+            var iterator = self.head;
+            while (iterator) |elem| : (iterator = elem.next) {
+                if (elem == elem_needle) return true;
+            }
+            return false;
+        }
+
+        /// Remove an element from the stack. Asserts that the element is in the stack.
+        /// This operation is O(N); if done frequently, consider a different data structure.
+        pub fn remove(self: *LIFO, to_remove: *T) void {
```

Comment:
Let's remove `remove`? And make `contains` private, for good measure. We can always generalize later, but better add one more speedbump on the path to accidentally O(N^2) something!